### ISSUE TITLE
help: black screen

### ISSUE THREAD

**Issue by gpyalt** [google/filament#1526]
This is not a bug, but a request for help. I am trying to integrate Filament into my graphics library, which has a test harness based on glut (I know, not very trendy). I rebuilt filament using vs2019/msvc, and I can verify it works using the samples. I also built `vk_hellotriangle` (which uses FilamentApp) and it works fine.

So I moved essentially the same code from `vk_hellotriangle.cpp` to my graphics library (which does OpenGL calls) and linked with filament. When I try to render via Filament, I can step through the code and all seems to work fine, but nothing is rendered on the screen, I don't even see a swapbuffer being done. 

I have verified I am passing the correct glut-created window to `engine->createSwapChain()`. but with the backend/driver architecture it is hard to follow exactly what happens. 

Any suggestion on what could be wrong, or how I could debug my issue?

Here is my code:

```
void xFilamentMgr::Create(uintptr_t window)
{
    _window = window;

    // Create Engine and Renderer
    // --------------------------
    _engine    = filament::Engine::create(filament::Engine::Backend::OPENGL);
    _swapChain = _engine->createSwapChain((void *)window);
    _renderer  = _engine->createRenderer();
    
    // create Scene, Camera and View
    // -----------------------------
    _scene     = _engine->createScene();
    _camera    = _engine->createCamera();
    _view      = _engine->createView();
    _view->setScene(_scene);
    _view->setCamera(_camera);
    _view->setVisibleLayers(0x4, 0x4);
    _view->setClearTargets(false, false, false);
    _view->setRenderTarget(View::TargetBufferFlags::DEPTH_AND_STENCIL);
    _view->setShadowsEnabled(false);
}
// ----------------------------------------------------------------
// ----------------------------------------------------------------
void xFilamentMgr::Draw()
{
    struct Vertex {
        filament::math::float2 position;
        uint32_t color;
    };

    static const Vertex TRIANGLE_VERTICES[3] = {
        {{1.0f, 0.0f}, 0xffff0000u},
        {{cos(M_PI * 2 / 3), sin(M_PI * 2 / 3)}, 0xff00ff00u},
        {{cos(M_PI * 4 / 3), sin(M_PI * 4 / 3)}, 0xff0000ffu},
    };

    static constexpr uint16_t TRIANGLE_INDICES[3] = { 0, 1, 2 };

    static bool init = false;
    static VertexBuffer* vb;
    static IndexBuffer* ib;
    static Material* mat;
    static Camera* cam;
    static Entity renderable;

    if (!init)
    {
        init = true;
        
        _view->setClearColor({0.1f, 0.125f, 0.25f, 1.0f});
        _view->setPostProcessingEnabled(false);
        _view->setDepthPrepass(filament::View::DepthPrepass::DISABLED);
        static_assert(sizeof(Vertex) == 12, "Strange vertex size.");
        vb = VertexBuffer::Builder()
            .vertexCount(3)
            .bufferCount(1)
            .attribute(VertexAttribute::POSITION, 0, VertexBuffer::AttributeType::FLOAT2, 0, 12)
            .attribute(VertexAttribute::COLOR, 0, VertexBuffer::AttributeType::UBYTE4, 8, 12)
            .normalized(VertexAttribute::COLOR)
            .build(*_engine);
        vb->setBufferAt(*_engine, 0,
                        VertexBuffer::BufferDescriptor(TRIANGLE_VERTICES, 36, nullptr));
        ib = IndexBuffer::Builder()
            .indexCount(3)
            .bufferType(IndexBuffer::IndexType::USHORT)
            .build(*_engine);
        ib->setBuffer(*_engine,
                      IndexBuffer::BufferDescriptor(TRIANGLE_INDICES, 6, nullptr));
        mat = Material::Builder()
            .package(RESOURCES_BAKEDCOLOR_DATA, RESOURCES_BAKEDCOLOR_SIZE)
            .build(*_engine);
        renderable = EntityManager::get().create();
        RenderableManager::Builder(1)
            .boundingBox({{ -1, -1, -1 }, { 1, 1, 1 }})
            .material(0, mat->getDefaultInstance())
            .geometry(0, RenderableManager::PrimitiveType::TRIANGLES, vb, ib, 0, 3)
            .culling(false)
            .receiveShadows(false)
            .castShadows(false)
            .build(*_engine, renderable);
        _scene->addEntity(renderable);
        cam = _engine->createCamera();
        _view->setCamera(cam);
    }
    
    RECT r;
    GetClientRect((HWND)_window, &r);
    _view->setViewport(Viewport(0, 0, r.right - r.left, r.bottom - r.top));

    static double now = 1;
    now += .01;
    constexpr float ZOOM = 1.5f;
    const uint32_t w = _view->getViewport().width;
    const uint32_t h = _view->getViewport().height;
    const float aspect = (float) w / h;
    cam->setProjection(Camera::Projection::ORTHO,
                       -aspect * ZOOM, aspect * ZOOM,
                       -ZOOM, ZOOM, 0, 1);
    auto& tcm = _engine->getTransformManager();
    tcm.setTransform(tcm.getInstance(renderable),
                     filament::math::mat4f::rotation(now, filament::math::float3{ 0, 0, 1 }));
    if (_renderer->beginFrame(_swapChain)) {
        _renderer->render(_view);
        _renderer->endFrame();
    }
}
```

**Comment by prideout** [comment]
Can you remove the following lines and try again?  I think you may've copied them from our Imgui layer, they don't make sense for the main view.

```
_view->setVisibleLayers(0x4, 0x4);
_view->setClearTargets(false, false, false);
_view->setRenderTarget(View::TargetBufferFlags::DEPTH_AND_STENCIL);
_view->setShadowsEnabled(false);
```

If you still see black instead of the clear color (which looks like it should be 0.1f, 0.125f, 0.25f, 1.0f in your case) then can you double check that Filament is making draw calls?  (i.e. put a breakpoint at `OpenGLDriver::draw`)

**Comment by gpyalt** [comment]
Thank you so much for the suggestions, Philip. I removed the 4 lines => no difference.

I set a breakpoints in OpenGLDriver::draw() and it is called:

![image](https://user-images.githubusercontent.com/40033896/63127291-f72a3100-bf7f-11e9-8216-4af181b165e7.png)

I also set breakpoints in PlatformWGL.cp and both `PlatformWGL::makeCurrent` and `PlatformWGL::commit` are called and execute `wglMakeCurrent(hdc, mContext);` and `SwapBuffers(hdc);`.

However, nothing happens in the window. Previous content (drawn previously without filament) remains, and if I resize the window I get black borders.

**Comment by prideout** [comment]
You mentioned that your graphics library makes OpenGL calls on its own?  That might be problematic since Filament creates its own GL context bound to the window handle that you pass in to the swap chain.

Another thing to try is enabling GL errors by either building Filament in a debug config or by removing lines 33 thru 39 in `GLUtils.h`

**Comment by gpyalt** [comment]
> You mentioned that your graphics library makes OpenGL calls on its own? 

yes, it does. Thanks for the pointer, I'll look into it further!

**Comment by gpyalt** [comment]
Hi @prideout, I got it working, thanks a lot for the help. 

The wglSetCurrent failed in PlatformWGL.cpp, I guess because it was creating a context for a pixel format maybe not compatible with the glut window. After a lot of poking around, I replaced the `CreateWindowA("STATIC", "dummy", 0, 0, 0, 1, 1, NULL, NULL, NULL, NULL)` with `FindWindow("FREEGLUT", "xtView")` and it started to work! Exciting times!

Again, many thanks for the help. You guys are fantastic.
